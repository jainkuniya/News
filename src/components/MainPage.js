//Import libraries
import React, { Component } from 'react';
import { View, Text, Picker, StyleSheet } from 'react-native';
import axios from 'axios';
import NewsList from './NewsList';
import styles from '../common/styles';
import { getSourcesUrl, getArticlesUrl } from '../networking/URL';
import SortBySelector from './SortBySelector';

//more styles
const moreStyles = StyleSheet.create({
	pickerStyle: {
		flex: 2
	},
	sourceSelector: {
		marginTop: 8,
		paddingLeft: 8,
		paddingRight: 8,
		flexDirection: 'row',
	},
	sourceSelectorText: {
		flex: 1,
		fontSize: 16,
	},
});

//Create class component
export default class MainPage extends Component {

	constructor() {
		super();
		this.state = {
			news: [],
			currentSourceId: 'the-next-web',
			sources: [],
			sortBy: 'latest',
			message: 'loading...',
		};
	}

	componentWillMount() {
		// Make a GET request to get news
		console.log(getSourcesUrl());
		axios.get(getSourcesUrl())
			.then(response => this.setState({ sources: response.data.sources }))
			.catch((error) => {
			console.log(error);
		});

		//fetch news with default source 
		this.fetchNews(this.state.currentSourceId, this.state.sortBy);
	}

	onValueChange = (id: string) => {
		//update selected source in picker
		this.setState({ 
			currentSourceId: id,
			news: [],
			message: 'loading...',
		});
		this.fetchNews(id, this.state.sortBy);		
	};

	onSortSelect = (sort: string) => {
		this.setState({
			sortBy: sort,
			news: [],
			message: 'loading...',
		});
		this.fetchNews(this.state.currentSourceId, sort);
	}

	fetchNews = (sourceId: string, sortBy: string) => {
		// Make a GET request to get news
		axios.get(getArticlesUrl(sourceId, sortBy))
			.then(response => this.setState({ news: response.data.articles }))
			.catch((error) => {
			console.log(error);
			this.setState({ 
				news: [],
				message: 'No news available :(',
			});
		});
	}

	renderSources() {
		return this.state.sources.map(source => 
			<Picker.Item 
				key={source.id} 
				label={source.name} 
				value={source.id}
			/>);
	}
	
	render() {
		const { pickerStyle, sourceSelector, sourceSelectorText } = moreStyles;

		return (
			<View>
				<View style={[sourceSelector, styles.cardView]}>
					<Text style={sourceSelectorText}>Sources</Text>
					<Picker
						style={[pickerStyle]}
						selectedValue={this.state.currentSourceId}
						onValueChange={(value) => this.onValueChange(value)}
					>
					{this.renderSources()}
					</Picker>
				</View>
				<View style={[sourceSelector, styles.cardView]}>
					<SortBySelector
						onSelect={this.onSortSelect}
						currentSelected={this.state.sortBy}
					/>
				</View>
				<NewsList
					newsList={this.state.news}
					message={this.state.message}
				/>
			</View>
		);
	}
}
