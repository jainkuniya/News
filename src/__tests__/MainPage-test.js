import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import MainPage from '../components/MainPage';

test('renders correctly', () => {
  const tree = renderer.create(
    <MainPage />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
