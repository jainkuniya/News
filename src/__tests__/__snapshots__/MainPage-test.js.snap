// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "marginTop": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
        },
        Object {
          "alignItems": "center",
          "backgroundColor": "#F8F8F8",
          "elevation": 2,
          "justifyContent": "center",
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.2,
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "flex": 1,
          "fontSize": 16,
        }
      }
    >
      Sources
    </Text>
    <View
      style={
        Array [
          Object {
            "flex": 2,
          },
        ]
      }
    >
      <RCTPicker
        items={Array []}
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        selectedIndex={0}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "marginTop": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
        },
        Object {
          "alignItems": "center",
          "backgroundColor": "#F8F8F8",
          "elevation": 2,
          "justifyContent": "center",
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.2,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "padding": 8,
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        isTVSelectable={true}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "borderColor": "#000000",
            "borderRadius": 4,
            "borderWidth": 0.5,
            "marginLeft": 16,
            "marginRight": 16,
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "paddingLeft": 16,
              "paddingRight": 16,
            }
          }
        >
          Latest
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        isTVSelectable={true}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "borderColor": "#d6d7da",
            "borderRadius": 4,
            "borderWidth": 0.5,
            "marginLeft": 16,
            "marginRight": 16,
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "paddingLeft": 16,
              "paddingRight": 16,
            }
          }
        >
          Top
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        isTVSelectable={true}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "borderColor": "#d6d7da",
            "borderRadius": 4,
            "borderWidth": 0.5,
            "marginLeft": 16,
            "marginRight": 16,
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "paddingLeft": 16,
              "paddingRight": 16,
            }
          }
        >
          Popular
        </Text>
      </View>
    </View>
  </View>
  <View>
    <RCTScrollView>
      <View />
    </RCTScrollView>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "paddingTop": 84,
          "position": "relative",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        loading...
      </Text>
    </View>
  </View>
</View>
`;
